util/build/sgx/src/libraries.rs

```diff
@@ -61,3 +81,59 @@ impl SgxLibraryCollection for [Library] {
         }
     }
 }
+
+/// Provides the necessary linker flags to link SGX libraries to the crate being
+/// built.
+///
+/// # Arguments
+///
+/// * `sgx` - The SGX environment for the current build.  This helps to
+///   determine if one needs to link to sim or hw libraries.
+pub fn link_to_sgx_libraries(sgx: &SgxEnvironment) -> Result<(), Error> {
+    let mut config = Config::new();
+    config
+        .exactly_version(SGX_VERSION)
+        .print_system_libs(true)
+        .cargo_metadata(false)
+        .env_metadata(true);
+    let lib_paths = if sgx.sgx_mode() == SgxMode::Simulation {
+        SGX_SIMULATION_LIBS
+    } else {
+        SGX_LIBS
+    }
+    .iter()
+    .map(|libname| Ok(config.probe(libname)?.link_paths))
+    .collect::<Result<Vec<Vec<PathBuf>>, PkgConfigError>>()?
+    .into_iter()
+    .flatten()
+    .collect::<HashSet<PathBuf>>();
```

Comment:
I'm still spinning up on rust iterators so any correction here is very welcome.  I basically cargo culted this from the prior implementation so this might be more for me to explain what's happening.

A good explanation that I started from, https://stackoverflow.com/a/59852696

The current implementation is collecting a result and then propagating an error out if there is one encountered.  If there is no error, the nested vector is flattened into a `HashSet`.  Per the [collect()](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect) docs:

> an iterator of Result<T, E> items can be collected into Result<Collection<T>, E>

The suggested implementation will use `flat_map` on the `Result`s. Which will have the effect of ignoring the errors, [IntoIterator for Result](https://doc.rust-lang.org/stable/core/result/enum.Result.html#impl-IntoIterator).

> The iterator yields one value if the result is Result::Ok, otherwise none.

The change to use `flat_map` would need to be along the lines of:
```
    .map(|libname| Ok(config.probe(libname)?.link_paths))
    .flat_map(|val| match val {
        Ok(vec) => vec.into_iter().map(Ok).collect(),
        Err(err) => vec![Err(err)],
    })
    .collect::<Result<HashSet<PathBuf>, PkgConfigError>>()?;
```

I'm not sure it's any better so will leave as is for now.

Reply:
I see. Maybe add a comment clarifying this intent? That stackoverflow link seems like a good summary.

Reply:
Thoughts on this comment
```
    // We want to propagate any error from `probe()`. This means that we must
    // collect the Results prior to iterating on them, since a IntoIterator for
    // Result<R, E> will omit any errors.
    // https://stackoverflow.com/questions/59852161/how-to-handle-result-in-flat-map/59852696#59852696
```

Another option is to pull in [itertools::flatten_ok](https://stackoverflow.com/a/69252224), that might omit the desire for a comment and it looks like itertools is already a transitive dependency.

Reply:
That comment SGTM.

`flatten_ok` also seems helpful, and would still benefit from a (potentially shorter) comment IMO, e.g. `NB: flat_map doesn't work here because Result::IntoIterator omits Errors`

Reply:
pushed up change that brings in `flatten_ok` and utilizes the comment as suggested.